EIPS/eip-2585.md

```diff
@@ -0,0 +1,443 @@
+ * and related and neighboring rights to this software to the public domain
+ * worldwide. This software is distributed without any warranty.
+ *
+ * You should have received a copy of the CC0 Public Domain Dedication along
+ * with this software. If not, see
+ * <https://creativecommons.org/publicdomain/zero/1.0/>.
+ *    
+ *       .-''-.  .-./`) .-------.               .`````-.  ,--------.     .-''''-.  ,--------.   
+ *     .'_ _   \ \ .-.')\  _(`)_ \             /   ,-.  \ |   _____|    /  _--.  \ |   _____|   
+ *    / ( ` )   '/ `-' \| (_ o._)|            (___/  |   ||  )          |_( )_ ' | |  )         
+ *   . (_ o _)  | `-'`"`|  (_,_) /_ _    _ _        .'  / |  '----.     (_ o _). / |  '----.    
+ *   |  (_,_)___| .---. |   '-.-'( ' )--( ' )   _.-'_.-'  |_.._ _  '.  .'(_,_).  `.|_.._ _  '.  
+ *   '  \   .---. |   | |   |   (_{;}_)(_{;}_)_/_  .'        ( ' )   \|_( )_    \  |  ( ' )   \ 
+ *    \  `-'    / |   | |   |    (_,_)--(_,_)( ' )(__..--. _(_{;}_)  |(_ o _)   /  |_(_{;}_)  | 
+ *     \       /  |   | /   )               (_{;}_)      ||  (_,_)  /  (_,_)..-' .'|  (_,_)  /  
+ *      `'-..-'   '---' `---'                (_,_)-------' `...__..'     `-....-'   `...__..'   
+ *                                                                                           
+ */
+pragma solidity 0.6.4;
+pragma experimental ABIEncoderV2;
+
+interface ERC1271 {
+    function isValidSignature(bytes calldata data, bytes calldata signature) external view returns (bytes4 magicValue);
+}
+
+interface ERC1654 {
+   function isValidSignature(bytes32 hash, bytes calldata signature) external view returns (bytes4 magicValue);
+}
+
+interface ReplayProtection {
+    function checkAndUpdateNonce(address signer, bytes calldata nonce) external returns (bool);
+}
+
+interface Forwarder {
+
+    enum SignatureType { DIRECT, EIP1654, EIP1271 }
+
+    struct Message {
+        address from;
+        address to;
+        uint256 chainId;
+        address replayProtection;
+        bytes nonce;
+        bytes data;
+        bytes32 innerMessageHash;
+	}
+
+    function forward(
+        Message calldata message,
+        SignatureType signatureType,
+        bytes calldata signature
+    ) external payable;
+}
+
+library SigUtil {
+    function recover(bytes32 hash, bytes memory sig) internal pure returns (address recovered) {
+        require(sig.length == 65, "SIGNATURE_INVALID_LENGTH");
+
+        bytes32 r;
+        bytes32 s;
+        uint8 v;
+        assembly {
+            r := mload(add(sig, 32))
+            s := mload(add(sig, 64))
+            v := byte(0, mload(add(sig, 96)))
+        }
+
+        // Version of signature should be 27 or 28, but 0 and 1 are also possible versions
+        if (v < 27) {
+            v += 27;
+        }
+        require(v == 27 || v == 28, "SIGNATURE_INVALID_V");
+
+        recovered = ecrecover(hash, v, r, s);
+        require(recovered != address(0), "SIGNATURE_ZERO_ADDRESS");
```

Comment:
is that check really necessary ? are there cases where ecrecover returns address(0) other then the signature actually matched that of a PK that correspond the address(0) ?